### PR TITLE
CBG-553: Split _sync xattr and doc key constants

### PR DESCRIPTION
**Pull request by JRascagneres**

### CODE REVIEW

File: db/design_doc.go [hunk 1]
```diff
@@ -90,15 +90,15 @@ const (
 	function(doc, meta) {
 	
 		//Skip any internal sync documents
-		if (meta.id.substring(0, 6) == "_sync:") {
+		if (meta.id.substring(0, 6) == "%s") {
```
Comment (bbrks) [inline db/design_doc.go]: I agree with parameterising or pulling these out as constants, but the long string of repeated `base.SyncPrefix, base.SyncXattrName, base.SyncPropertyName, base.SyncXattrName, base.SyncPropertyName` in the Sprintf isn't ideal. Makes it difficult to determine where to insert new values if we need to change this.

Can we either do string concatenation inline using the constants, or maybe positional arguments? I can't decide which is better...

E.g: for inline concatenation: ```(meta.id.substring(0, 6) == "`+base.SyncPrefix+`") {```
Or Sprintf positional args https://play.golang.org/p/NMVLgVO0y0v

File: db/design_doc.go [hunk 2]
```diff
@@ -348,7 +348,7 @@ func installViews(bucket base.Bucket) error {
 		                  	} else {
 		                       	sync = meta.xattrs.%s
 		                    }
-		                     `, base.SyncXattrName, base.SyncXattrName)
+		                     `, base.SyncPropertyName, base.SyncXattrName)
```
Comment (bbrks) [inline db/design_doc.go]: Looks like both of these should be xattr, but we need to parameterise the `doc._sync` as well.

**Review by bbrks — CHANGES_REQUESTED** [review]

File: db/design_doc.go (hunk 1 above)
Comment (JRascagneres) [reply db/design_doc.go]: Gone with positional args. I prefer the look of that.

**Review by bbrks — APPROVED** [review]
LGTM 👍

**Comment by adamcfraser** [comment]
Looks good - can you do a quick cleanup of any hardcoded usages of "_sync" in our code.

**Comment by JRascagneres** [comment]
@adamcfraser Replaced _sync constants. There ended up adding quite a few so we can see what you think.

File: base/bucket_gocb_test.go [hunk 3]
```diff
@@ -2222,9 +2221,9 @@ func createTombstonedDoc(bucket *CouchbaseBucketGoCB, key, xattrName string) {
 
 	// Create tombstone revision which deletes doc body but preserves XATTR
 	_, mutateErr := bucket.Bucket.MutateInEx(key, flags, gocb.Cas(cas), uint32(0)).
-		UpsertEx(xattrName, xattrVal, gocb.SubdocFlagXattr).                                     // Update the xattr
-		UpsertEx("_sync.cas", "${Mutation.CAS}", gocb.SubdocFlagXattr|gocb.SubdocFlagUseMacros). // Stamp the cas on the xattr
-		RemoveEx("", gocb.SubdocFlagNone).                                                       // Delete the document body
+		UpsertEx(xattrName, xattrVal, gocb.SubdocFlagXattr).                                                 // Update the xattr
+		UpsertEx(SyncPropertyName+".cas", "${Mutation.CAS}", gocb.SubdocFlagXattr|gocb.SubdocFlagUseMacros). // Stamp the cas on the xattr
```
Comment (adamcfraser) [inline base/bucket_gocb_test.go]: It looks like this should be xattrName instead of SyncPropertyName.

File: db/change_cache.go [hunk 4]
```diff
@@ -472,7 +472,7 @@ func (c *changeCache) DocChanged(event sgbucket.FeedEvent) {
 
 	// If the doc update wasted any sequences due to conflicts, add empty entries for them:
 	for _, seq := range syncData.UnusedSequences {
-		base.Infof(base.KeyCache, "Received unused #%d in _sync.unused_sequences property for (%q / %q)", seq, base.UD(docID), syncData.CurrentRev)
+		base.Infof(base.KeyCache, "Received unused #%d in "+base.SyncPropertyName+".unused_sequences property for (%q / %q)", seq, base.UD(docID), syncData.CurrentRev)
```
Comment (adamcfraser) [inline db/change_cache.go]: For this case and the next one, I think the _sync could be removed altogether.
```suggestion
		base.Infof(base.KeyCache, "Received unused #%d in unused_sequences property for (%q / %q)", seq, base.UD(docID), syncData.CurrentRev)
```

File: db/change_cache.go [hunk 5]
```diff
@@ -495,7 +495,7 @@ func (c *changeCache) DocChanged(event sgbucket.FeedEvent) {
 
 		for _, seq := range syncData.RecentSequences {
 			if seq >= c.getNextSequence() && seq < currentSequence {
-				base.Infof(base.KeyCache, "Received deduplicated #%d in _sync.recent_sequences property for (%q / %q)", seq, base.UD(docID), syncData.CurrentRev)
+				base.Infof(base.KeyCache, "Received deduplicated #%d in "+base.SyncPropertyName+".recent_sequences property for (%q / %q)", seq, base.UD(docID), syncData.CurrentRev)
```
Comment (adamcfraser) [inline db/change_cache.go]:
```suggestion
				base.Infof(base.KeyCache, "Received deduplicated #%d in recent_sequences property for (%q / %q)", seq, base.UD(docID), syncData.CurrentRev)
```

File: db/crud_test.go [hunk 6]
```diff
@@ -157,7 +157,7 @@ func TestRevisionStorageConflictAndTombstones(t *testing.T) {
 	goassert.DeepEquals(t, gotbody, rev2a_body)
 
 	// Ensure previous revision body backup has been removed
-	_, _, err = db.Bucket.GetRaw("_sync:rb:4GctXhLVg13d59D0PUTPRD0i58Hbe1d0djgo1qOEpfI=")
+	_, _, err = db.Bucket.GetRaw(base.SyncPrefix + "rb:4GctXhLVg13d59D0PUTPRD0i58Hbe1d0djgo1qOEpfI=")
```
Comment (adamcfraser) [inline db/crud_test.go]: This and the cases below are base.RevBodyPrefix.
```suggestion
	_, _, err = db.Bucket.GetRaw(base.RevBodyPrefix + "4GctXhLVg13d59D0PUTPRD0i58Hbe1d0djgo1qOEpfI=")
```

File: db/indexes.go [hunk 7]
```diff
@@ -26,9 +26,9 @@ const (
 // When running with xattrs, that gets replaced with META().xattrs._sync (or META(bucketname).xattrs._sync for query).
 // When running w/out xattrs, it's just replaced by the doc path `bucketname`._sync
 // This gets replaced before the statement is sent to N1QL by the replaceSyncTokens methods.
-var syncNoXattr = fmt.Sprintf("`%s`._sync", base.BucketQueryToken)
-var syncXattr = "meta().xattrs._sync"
-var syncXattrQuery = fmt.Sprintf("meta(`%s`).xattrs._sync", base.BucketQueryToken) // Replacement for $sync token for xattr queries
+var syncNoXattr = fmt.Sprintf("`%s`.%s", base.BucketQueryToken, base.SyncPropertyName)
+var syncXattr = "meta().xattrs." + base.SyncXattrName
+var syncXattrQuery = fmt.Sprintf("meta(`%s`).xattrs."+base.SyncXattrName, base.BucketQueryToken) // Replacement for $sync token for xattr queries
```
Comment (adamcfraser) [inline db/indexes.go]: Another case of mixed printf+concatenation.

File: rest/blip_api_test.go [hunk 8]
```diff
@@ -1455,7 +1455,7 @@ func TestGetRemovedDoc(t *testing.T) {
 	rt.GetDatabase().FlushRevisionCacheForTest()
 
 	// Delete any temp revisions in case this prevents the bug from showing up (didn't make a difference)
-	tempRevisionDocId := "_sync:rev:foo:5:3-cde"
+	tempRevisionDocId := base.SyncPrefix + "rev:foo:5:3-cde"
```
Comment (adamcfraser) [inline rest/blip_api_test.go]: base.RevPrefix here.

**Review by adamcfraser — CHANGES_REQUESTED** [review]

**Review by adamcfraser — APPROVED** [review]

**Pull request merged**